### PR TITLE
Refactored ConvL_ConcRT.py, ResponseTime.py and shifted them to channel.py

### PR DESCRIPTION
**Pull request by krishnacharya**
## What? Why?
Changes proposed in this pull request:
Re-factored ConvL_ConcRT.py to work with new implementation of reader.py and shifted it to channel.py
Re-factored ResponseTIme.py to work with new implementation of reader.py and shifted it to channel.py

also added constants required for conv_len_conv_refr_time functioi in channel.py into config.py , similarly added constants into config.py for find_response_time function in channel.py
made changes to Tester.py to test re-factored code
fixes bug in saver.py

## Checks
- [x] Single commit and [No merge commits](http://nathanleclaire.com/blog/2014/09/14/dont-be-scared-of-git-rebase/)
- [x] Make sure you have added required documentation
- [x] If you have refactored a code, make sure that you have compared the outputs pre and post refactoring. Add both pre and post refaoring screenshots below if the output is visual

## Any images?
previous Conversation length:
[#kubuntu-devel_1-1_1-31_CL.xlsx](https://github.com/prasadtalasila/IRCLogParser/files/772371/kubuntu-devel_1-1_1-31_CL.xlsx)

previous Conversation Refresh Time:
[#kubuntu-devel_1-1_1-31_CRT.xlsx](https://github.com/prasadtalasila/IRCLogParser/files/772372/kubuntu-devel_1-1_1-31_CRT.xlsx)

new  Conversation length:
[CL.xlsx](https://github.com/prasadtalasila/IRCLogParser/files/772374/CL.xlsx)

new Conversation Refresh Time:
[CRT.xlsx](https://github.com/prasadtalasila/IRCLogParser/files/772375/CRT.xlsx)

previous Response Time:
[#kubuntu-devel_1-1_1-31_RT.xlsx](https://github.com/prasadtalasila/IRCLogParser/files/773260/kubuntu-devel_1-1_1-31_RT.xlsx)

new Response Time:
[RT.xlsx](https://github.com/prasadtalasila/IRCLogParser/files/773261/RT.xlsx)

## Notify reviewers
@rohangoel96

### CODE REVIEW

File: lib/analysis/ConvL_ConvRT.py [hunk 1]
```diff
@@ -1,147 +1,44 @@
...
-				break
-			if not nick_same_list[ind]:
-				nick_same_list[ind].append(ni)
-				break
-
-	G = ext.util.to_graph(nick_same_list)
+	G = util.to_graph(nick_same_list)
 	L = list(connected_components(G))
 
 	for i in range(1,len(L)+1):
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: for i in range(len(L))

File: lib/analysis/ConvL_ConvRT.py [hunk 2]
```diff
@@ -1,147 +1,44 @@
...
-                L[i-1] = list(L[i-1])
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: use i instead of i-1

File: lib/analysis/ConvL_ConvRT.py [hunk 3]
```diff
@@ -1,147 +1,44 @@
...
+		L[i-1] = list(L[i-1])
 
 	# We use connected components algorithm to group all those nick clusters that have atleast one nick common in their clusters. So e.g.
 	#Cluster 1- nick1,nick2,nick3,nick4(some nicks of a user) #Cluster 2 -nick5,nick6,nick2,nick7. Then we would get - nick1,nick2,nick3,nick4,nick5,nick6,nick7 and we can safely assume they belong to the same user.
 
-	conversations=[[] for i in range(10000)] #This might need to be incremented from 10000 if we have more users. Same logic as the above 7000 one. Applies to all the other codes too.
-	graph_to_sir = []                ## I would advice on using a different data structure which does not have an upper bound like we do in arrays. 
+	conversations=[[] for i in range(config.MAX_CONVERSATIONS)] #This might need to be incremented from 10000 if we have more users. Same logic as the above 7000 one. Applies to all the other codes too.
+	graph_to_sir = []                                           ## I would advice on using a different data structure which does not have an upper bound like we do in arrays.
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: better variable names?

File: lib/analysis/ConvL_ConvRT.py [hunk 4]
```diff
@@ -150,36 +47,27 @@ def findConvLength_ConvRefreshTime(log_directory, channel_name, output_directory
...
-				if not((folderiterator==2 and (fileiterator ==29 or fileiterator ==30 or fileiterator ==31)) or ((folderiterator==4 or folderiterator==6 or folderiterator==9 or folderiterator==11) and fileiterator==31 )): 
-					print "[Error] Path "+filePath+" doesn't exist"
-				continue 
-			with open(filePath) as f:
-							content = f.readlines() #contents stores all the lines of the file channel_name  
+	for day_content_all_channels in log_dict.values():
+		for day_content in day_content_all_channels:
+			day_log = day_content["log_data"]
+
 			dateadd=dateadd+1
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: spaces around = ?

File: lib/analysis/ConvL_ConvRT.py [hunk 5]
```diff
@@ -229,20 +117,20 @@ def findConvLength_ConvRefreshTime(log_directory, channel_name, output_directory
...
 											if(len(conversations[rt])==0):
 												conversations[rt].append(nick_sender)
 												conversations[rt].append(nick_receiver)
-												conversations[rt].append(24*60*dateadd + int(line[1:6][0:2])*60+int(line[1:6][3:5]))
+												conversations[rt].append(config.HOURS_PER_DAY*config.MINS_PER_HOUR*dateadd + int(line[1:6][0:2])*config.MINS_PER_HOUR+int(line[1:6][3:5]))
 												break
 
 						if(flag_comma == 0):
 							rec=line[line.find(">")+1:line.find(", ")] 
 							rec=rec[1:]
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: can be combined with line above?

File: lib/analysis/ConvL_ConvRT.py [hunk 6]
```diff
@@ -1,147 +1,44 @@
...
-	for ni in nicks:
-		for ind in range(7000):
-			if ni in nick_same_list[ind]:
-				break
-			if not nick_same_list[ind]:
-				nick_same_list[ind].append(ni)
-				break
-
-	G = ext.util.to_graph(nick_same_list)
+	G = util.to_graph(nick_same_list)
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: better variable names? https://github.com/prasadtalasila/IRCLogParser/blob/dev/lib/analysis/network.py#L32

File: lib/analysis/ConvL_ConvRT.py [hunk 7]
```diff
@@ -150,36 +47,27 @@ def findConvLength_ConvRefreshTime(log_directory, channel_name, output_directory
...
+
 			dateadd=dateadd+1
 			send_time = [] #list of all the times a user sends a message to another user
 			meanstd_list = []
 			totalmeanstd_list = []
 			x_axis = []
 			y_axis = []
 			real_y_axis = []
-			time_in_min = [[] for i in range(1000)]
+			time_in_min = [[] for i in range(1000)]
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: why this constant?

File: lib/analysis/ConvL_ConvRT.py [hunk 8]
```diff
@@ -150,36 +47,27 @@ def findConvLength_ConvRefreshTime(log_directory, channel_name, output_directory
...
+			time_in_min = [[] for i in range(1000)]	
 			
-			print(filePath)
-
 			#code for making relation map between clients		
-			for line in content:
+			for line in day_log:
 				flag_comma = 0
 				if(line[0] != '=' and "] <" in line and "> " in line):
 					m = re.search(r"\<(.*?)\>", line)
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: line 67 to 69 can vbe combined in one line

File: lib/analysis/ConvL_ConvRT.py [hunk 9]
```diff
@@ -192,7 +80,7 @@ def findConvLength_ConvRefreshTime(log_directory, channel_name, output_directory
 							break
 						for ik in xrange(0,len(rec_list)):
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: better variable names?

File: lib/analysis/ConvL_ConvRT.py [hunk 10]
```diff
@@ -293,140 +181,6 @@ def findConvLength_ConvRefreshTime(log_directory, channel_name, output_directory
...
-
-rows = zip(graph_x_axis,graph_y_axis)         #Storing the RT values and their frequencies in csv file.
-with open('/home/dhruvie/LOP/graphforsir2.csv', 'a+') as myfile:
-				wr = csv.writer(myfile, quoting=csv.QUOTE_ALL)
-				for row in rows:
-					wr.writerow(row)
-
-
-'''
+	return row_cl,row_crt
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: space after comma?

File: lib/analysis/ConvL_ConvRT.py [hunk 11]
```diff
@@ -1,147 +1,44 @@
...
+import networkx as nx
+import re
+from networkx.algorithms.components.connected import connected_components
 import sys
 import csv
-import ext.util
+import util
+import config
+
+def findConvLength_ConvRefreshTime(log_dict, nicks,nick_same_list):
```
Comment (rohangoel96) [inline lib/analysis/ConvL_ConvRT.py]: dont use camel case.
function should be called conv_len_conv_refr_time()

File: TestCLRT.py [hunk 12]
```diff
@@ -0,0 +1,19 @@
+import sys
```
Comment (rohangoel96) [inline TestCLRT.py]: We will use only one Tester.py file. Kindly test in that only.
Don't create new  tester files.

Comment (prasadtalasila) [reply TestCLRT.py]: Is having one test file a standard practice in python? In TDD workflow, we typically have one or more test files for each code file.

File: lib/analysis/channel.py [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+		row_crt(zip list) :Conversation Refresh time
+	   
+	"""	
+	conv = []
+	conv_diff = []	
+
+	G = util.to_graph(nick_same_list)
+	conn_comp_list = list(connected_components(G))
+
+	for i in range(0,len(conn_comp_list)):
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: no need of 0

File: lib/analysis/channel.py [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+					nick_to_search = util.correctLastCharCR(m.group(0)[1:-1])
+					for d in range(len(nicks)):                        #E.g. if names are rohan1,rohan2,rohan3...,then var will store rohan1.
+						if((d < len(conn_comp_list)) and (nick_to_search in conn_comp_list[d])):
+							nick_sender = conn_comp_list[d][0]
+							break
+						
+					for i in nicks:
+						rec_list = [e.strip() for e in line.split(':')]
+						rec_list[1] = rec_list[1][rec_list[1].find(">") + 1:len(rec_list[1])]
+						rec_list[1] = rec_list[1][1:]
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: combine with line above and do it in one line

File: lib/analysis/channel.py [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+						if not rec_list[1]:
+							break
+						for ik in xrange(0,len(rec_list)):
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: ik ? use better variable names

File: lib/analysis/channel.py [hunk 16]
```diff
@@ -0,0 +1,367 @@
...
+											break
+										if(len(conversations[rt]) == 0):
+											conversations[rt].append(nick_sender)
+											conversations[rt].append(nick_receiver)
+											conversations[rt].append(config.HOURS_PER_DAY*config.MINS_PER_HOUR*dateadd + int(line[1:6][0:2])*config.MINS_PER_HOUR + int(line[1:6][3:5]))
+											break
+		
+	#Lines 212-290 consider all cases in which messages are addressed as - (nick1:nick2 or nick1,nick2 or nick1,nick2:) and stores their response times in conversations. conversations[i] contains all the response times between userA and userB throughout an entire year.
+
+	for ty in range(0,len(conversations)):       #Lines 295-297 remove the first two elements from every conversations[i] as they are the UIDS of sender and receiver respectively(and not RTs)
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: no need of 0

File: lib/analysis/channel.py [hunk 17]
```diff
@@ -0,0 +1,367 @@
...
+		if(len(conversations[ty]) != 0):              # response times are calculated starting from index 2. So now we have all the response times in conversations.
+			del conversations[ty][0:2]
+
+	for fg in range(0,len(conversations)):
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: ty? fg? use better names

File: lib/analysis/channel.py [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+						first = conversations[fg][gh]
+					if(gh == (len(conversations[fg]) - 1)):
+						conv.append(conversations[fg][gh] - first)					
+						break
+
+	for op in range(0,max(conv)):
+		graphx1.append(op)
+		graphy1.append(conv.count(op))
+
+	for po in range(0,max(conv_diff)):
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: po?

File: lib/analysis/channel.py [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+		graphy2.append(conv_diff.count(po))
+
+#To plot CDF we store the CL and CRT values and their number of occurences as shown above.
+
+	row_cl = zip(graphx1,graphy1)
+	row_crt = zip(graphx2,graphy2)
+	
+	return row_cl, row_crt
+
+def find_response_time(log_dict, nicks, nick_same_list):
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: response_time()

File: lib/analysis/channel.py [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+	conv = []
+	conv_diff = []
+	
+	G = util.to_graph(nick_same_list)
+	conn_comp_list = list(connected_components(G))
+
+	for i in range(0,len(conn_comp_list)):
+		conn_comp_list[i] = list(conn_comp_list[i])
+	
+	graph_to_sir = []
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: graph to sir?

File: lib/analysis/channel.py [hunk 21]
```diff
@@ -0,0 +1,367 @@
...
+
+		#Explanation provided in parser-CL+CRT.py
+			for index in range(0,config.MAX_RESPONSE_CONVERSATIONS):
+				if(len(conversations[index]) != 0):
+					for index1 in range(2,len(conversations[index])):
+						totalmeanstd_list.append(conversations[index][index1])
+
+			if(len(totalmeanstd_list) != 0):
+				for iy in range(0, max(totalmeanstd_list) + 1):
+					x_axis.append(iy)
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: iy ? ui?

File: lib/analysis/channel.py [hunk 22]
```diff
@@ -0,0 +1,367 @@
...
+
+				for ui in x_axis:
+					y_axis.append(float(totalmeanstd_list.count(ui)) / float(len(totalmeanstd_list)))
+				
+				#finding the probability of each RT to occur=No. of occurence/total occurences.
+				real_y_axis.append(y_axis[0])
+				for ix in range(1, len(y_axis)):
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: ix? ih?

File: lib/analysis/channel.py [hunk 23]
```diff
@@ -0,0 +1,367 @@
...
+		
+			for index in range(0,config.MAX_RESPONSE_CONVERSATIONS):
+				if(len(conversations[index])!=0):
+					for index1 in range(2,len(conversations[index])):
+						meanstd_list.append(conversations[index][index1])
+					conversations[index].append(numpy.mean(meanstd_list))
+					conversations[index].append(numpy.mean(meanstd_list)+(2*numpy.std(meanstd_list)))
+					meanstd_list[:] = []
+
+	graph_to_sir.sort()
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: graph to sir?

File: Tester.py [hunk 24]
```diff
@@ -17,13 +17,21 @@
...
 
 #output
 # saver.draw_nx_graph(message_number_graph, output_directory, "message_time_graph")
-for i in range(len(nick_change_graph_list)):
-	saver.draw_nx_graph(nick_change_graph_list[i], output_directory, "nick_change_graph_" + str(i))
+#for i in range(len(nick_change_graph_list)):
+#	saver.draw_nx_graph(nick_change_graph_list[i], output_directory, "nick_change_graph_" + str(i))
 
+# Test for Conversation Length and Conversation Refresh Time
+CL,CRT = channel.conv_len_conv_refr_time(log_data,nicks,nick_same_list)
```
Comment (rohangoel96) [inline Tester.py]: use proper names. caps are for constants.
conv_len, conv_ref_time are better names

Comment (rohangoel96) [inline Tester.py]: use spaces after commas?

File: Tester.py [hunk 25]
```diff
@@ -17,13 +17,21 @@
...
+saver.save_csv(CL,output_directory,"CL")
+saver.save_csv(CRT,output_directory,"CRT")
+
+# Test for Response Time
+RT = channel.find_response_time(log_data,nicks,nick_same_list)
```
Comment (rohangoel96) [inline Tester.py]: This is analysis should be under the analysis comment and output under output comment.

File: lib/analysis/channel.py [hunk 26]
```diff
@@ -0,0 +1,367 @@
...
+
+	""" Calculates the conversation length (CL) that is the length of time for which two users communicate 
+	i.e. if a message is not replied to within Response Time(RT), 
+	then it is considered as a part of another conversation.
+	This function also calculates the conversation refresh time(CRT). 
+	For a pair of users, this is the time when one conversation ends and another one starts.
+	Args:
+		log_dict (str): Dictionary of logs data created using reader.py
+		nicks(list) : list of nickname created using nickTracker.py
+		nick_same_list :list of same_nick names created using nickTracker.py
```
Comment (rohangoel96) [inline lib/analysis/channel.py]: First letter should start with caps "List"

**Pull request merged**